dpnp/dpnp_algo/dpnp_elementwise_common.py

```diff
@@ -795,6 +826,331 @@ def __call__(self, *args, **kwargs):
+        x1_dt = _get_dtype(x1, sycl_dev)
+        x2_dt = _get_dtype(x2, sycl_dev)
+        if not all(
+            _validate_dtype(dt) for dt in [x1_dt, x2_dt]
+        ):  # pragma: no cover
+            raise ValueError("Operands have unsupported data types")
+
+        x1_dt, x2_dt = self.get_array_dtype_scalar_type_resolver_function()(
+            x1_dt, x2_dt, sycl_dev
+        )
+
+        buf1_dt, buf2_dt, res1_dt, res2_dt = find_buf_dtype_4out(
+            x1_dt,
+            x2_dt,
+            self.get_type_result_resolver_function(),
+            sycl_dev,
+        )
+        if res1_dt is None or res2_dt is None:
+            raise ValueError(
+                f"function '{self.name_}' does not support input type "
+                f"({x1_dt}, {x2_dt}), "
+                "and the input could not be safely coerced to any "
+                "supported types according to the casting rule ''safe''."
+            )
+        buf_dts = [buf1_dt, buf2_dt]
+
+        if not isinstance(out, tuple):
+            raise TypeError("'out' must be a tuple of arrays")
+
+        if len(out) != self.nout:
+            raise ValueError(
+                "'out' tuple must have exactly one entry per ufunc output"
+            )
+
+        if not (out1 is None and out2 is None):
+            if all(res is None for res in out):
+                out = (out1, out2)
+            else:
+                raise TypeError(
+                    "cannot specify 'out' as both a positional and keyword argument"
+                )
+
+        orig_out, out = list(out), list(out)
+        res_dts = [res1_dt, res2_dt]
+
+        for i in range(self.nout):
+            if out[i] is None:
+                continue
+
+            res = dpnp.get_usm_ndarray(out[i])
+            if not res.flags.writable:
+                raise ValueError("output array is read-only")
+
+            if res.shape != res_shape:
+                raise ValueError(
+                    "The shape of input and output arrays are inconsistent. "
+                    f"Expected output shape is {res_shape}, got {res.shape}"
+                )
+
+            if dpu.get_execution_queue((exec_q, res.sycl_queue)) is None:
+                raise dpnp.exceptions.ExecutionPlacementError(
+                    "Input and output allocation queues are not compatible"
+                )
+
+            res_dt = res_dts[i]
+            if res_dt != res.dtype:
+                if not dpnp.can_cast(res_dt, res.dtype, casting="same_kind"):
+                    raise TypeError(
+                        f"Cannot cast ufunc '{self.name_}' output {i + 1} from "
+                        f"{res_dt} to {res.dtype} with casting rule 'same_kind'"
+                    )
+
+                # Allocate a temporary buffer with the required dtype
+                out[i] = dpt.empty_like(res, dtype=res_dt)
+            else:
+                for x, dt in zip([x1, x2], buf_dts):
+                    if dpnp.isscalar(x):
```

Comment:
I do not quite understand the logic here
Regardless of these checks we always allocate a temporary buffer `out[i] = dpt.empty_like(res)`
What is the point of the checks?

Reply:
The point is to allocate a temporary buffer to avoid memory overlapping of any input array with any of passed `out` array.
We don't need to allocate a temporary, when:
* input array is a scalar, because none `out` can be a scalar
* `dt` is not None, because we will allocate a temporary buffer for `x` array
* input array `x` does not overlap any `out` array

Reply:
Rewritten.